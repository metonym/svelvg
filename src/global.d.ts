declare module "estree-walker" {
  import type { Element } from "svelte/types/compiler/interfaces";

  export function walk(
    ast: Ast,
    options: {
      enter: (node: Element, parentNode: Element) => void;
    },
  ): void;
}
