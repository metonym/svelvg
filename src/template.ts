import { parse, walk } from "svelte/compiler";

interface Position {
  start: number;
  end: number;
}

interface NodeElement extends Position {
  type: "Element";
  name: string;
  children: Position[];
  attributes: {
    name: string;
    value: { raw: any }[];
  }[];
}

export const templateSvelte = (
  source: string,
  filename: string,
  { classes }: { classes: string[] }
) => {
  const ast = parse(source, { filename });

  let svg_attributes = "";
  let svg_children = "";

  classes.forEach((name) => (svg_attributes += ` class:${name}={true}`));

  walk(ast.html, {
    enter(node: NodeElement) {
      if (node.type === "Element" && node.name === "svg") {
        node.children.forEach((child) => {
          svg_children += source.slice(child.start, child.end);
        });

        node.attributes.forEach(({ name, value }) => {
          if (name === "class") {
            value[0].raw.split(" ").forEach((name: string) => {
              svg_attributes += ` class:${name}={true}`;
            });
          } else {
            svg_attributes += ` ${name}="${value[0].raw}"`;
          }
        });
      }
    },
  });

  return `<svg${svg_attributes} {...$$restProps}><slot />${svg_children.trim()}</svg>`;
};

export const templateTs = (
  moduleName: string
) => `/// <reference types="svelte" />
import type { SvelteComponentTyped } from "svelte";

export interface ${moduleName}Props
  extends svelte.JSX.SVGAttributes<SVGSVGElement> {}

export default class ${moduleName} extends SvelteComponentTyped<
  ${moduleName}Props,
  {},
  { default: {} }
> {}
`;
