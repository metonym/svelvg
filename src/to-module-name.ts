export const toModuleName = (name: string) =>
  name
    .replace(/\-+/g, " ")
    .replace(/\.svg/g, "")
    .split(" ")
    .map((_, i) => {
      let first_char = _.slice(0, 1);

      if (i === 0 && first_char.match(/[0-9]|!/)) {
        first_char = "_" + first_char;
      }

      return first_char.toUpperCase() + _.slice(1);
    })
    .join("");
