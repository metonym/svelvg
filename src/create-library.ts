import tg from "tiny-glob";
import fs from "fs/promises";
import { existsSync } from "fs";
import path from "path";
import { templateSvelte, templateTs } from "./template";
import { toModuleName } from "./to-module-name";

interface CreateLibraryOptions {
  /** @default "lib" */
  outDir: string;

  /** @default false */
  iconIndex: boolean | string;

  /**
   * Callback to add a list of classes to the SVG element
   * provided the original filename and module name
   * @example
   * filename: "alarm-fill"
   * moduleName: "AlarmFill"
   */
  appendClassNames: (
    filename: string,
    moduleName: string
  ) => undefined | string[];

  /**
   * Override the default module name
   */
  toModuleName: (params: {
    path: path.ParsedPath;
    moduleName: string;
  }) => string;
}

type ModuleNames = string[];

export const createIndexFile = (moduleNames: ModuleNames) =>
  moduleNames
    .map(
      (moduleName) =>
        `export { default as ${moduleName} } from "./${moduleName}.svelte";\n`
    )
    .join("");

export const createIconIndex = (moduleNames: ModuleNames) => `# Icon Index

> ${moduleNames.length} total icons

## Icons

${moduleNames.map((moduleName) => `- ${moduleName}\n`).join("")}`;

export async function createLibrary(
  glob: string,
  options: Partial<CreateLibraryOptions>
) {
  const outDir = options?.outDir ?? "lib";

  let iconIndex: boolean | string = false;

  if (typeof options?.iconIndex === "string") {
    iconIndex = options.iconIndex;
  } else if (options?.iconIndex === true) {
    iconIndex = "ICON_INDEX.md";
  }

  let libPath = glob;

  if (!/^node_modules/.test(libPath)) {
    libPath = path.join("node_modules", glob);
  }

  if (!/\*.svg$/.test(libPath)) {
    libPath = path.join(libPath, "*.svg");
  }

  const libDir = path.parse(libPath).dir;

  try {
    await fs.stat(libDir);
  } catch (error) {
    console.error(`[createLibrary] path does not exist: ${libDir}`);
    process.exit(1);
  }

  try {
    const dir = path.join(process.cwd(), outDir);

    if (existsSync(dir)) {
      await fs.rm(dir, { recursive: true });
    }
    await fs.mkdir(dir);

    const files = await tg(libPath.replace(/\*.svg$/, "**/*.svg"));
    const moduleNames: ModuleNames = [];

    for (const filename of files) {
      const source = await fs.readFile(filename, "utf-8");
      const parsedPath = path.parse(filename);

      let moduleName = toModuleName(parsedPath.name);

      if (options?.toModuleName) {
        moduleName = options.toModuleName({ path: parsedPath, moduleName });
      }

      const classes =
        options?.appendClassNames?.call(null, parsedPath.name, moduleName) ??
        [];

      fs.writeFile(
        path.join(dir, `${moduleName}.svelte`),
        templateSvelte(source, filename, { classes })
      );
      fs.writeFile(
        path.join(dir, `${moduleName}.svelte.d.ts`),
        templateTs(moduleName)
      );

      moduleNames.push(moduleName);
    }

    const uniqueModuleNames = [...new Set(moduleNames)];
    const index = createIndexFile(uniqueModuleNames);

    fs.writeFile(path.join(dir, "index.js"), index);
    fs.writeFile(path.join(dir, "index.d.ts"), index);

    console.log(
      `⚡ Converted ${uniqueModuleNames.length} icons from "${glob}" to Svelte components in "${outDir}"`
    );

    if (iconIndex) {
      fs.writeFile(
        path.join(process.cwd(), iconIndex),
        createIconIndex(uniqueModuleNames)
      );

      console.log(`✏️ Wrote icon index to "${iconIndex}"`);
    }
  } catch (error) {
    console.error(error);
  }
}
