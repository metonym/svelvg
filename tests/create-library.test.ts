import { test, expect } from "vitest";
import { createIndexFile, createIconIndex } from "../src/create-library";

const moduleNames = ["OneTwoThree", "TwoThreeFour"];

test("createIndexFile", () => {
  expect(createIndexFile(moduleNames)).toMatchSnapshot();
});

test("createIconIndex", () => {
  expect(createIconIndex(moduleNames)).toMatchSnapshot();
});
