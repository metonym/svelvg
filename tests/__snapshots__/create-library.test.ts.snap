// Vitest Snapshot v1

exports[`createIconIndex 1`] = `
"# Icon Index

> 2 total icons

## Icons

- OneTwoThree
- TwoThreeFour

"
`;

exports[`createIndexFile 1`] = `
"export { default as OneTwoThree } from \\"./OneTwoThree.svelte\\";
export { default as TwoThreeFour } from \\"./TwoThreeFour.svelte\\";
"
`;
